### ISSUE TITLE
Helm Chart Allow for Empty data settings

### ISSUE THREAD

**Issue by jpyen** [apache/airflow#16684]
The helm-chart should not force the settings of Resultsbackend, Metadata and Broker. Either through an additional configuration option or implicitly by not providing values, the helm-chart should refrain from populating those Airflow-Env configurations.

**Use case / motivation**
The current Helm chart forces the configuration of certain values such as `brokerUrl` (https://github.com/apache/airflow/blob/main/chart/values.yaml#L236)
Which will then be populated as Airflow ENV config.
Due to Airflows configuration precedence (https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html) are these settings taking highest priority in Airflow configuration.

With the suggested change, users can dynamically provide Airflow configurations through Airflow's `__CMD` environment variables.

**In other words; provide users with the option to utilize Airflow's `__CMD` ENV configuration.**

_I am willing to submit a PR._

**Comment by potiuk** [comment]
I think this is actually very good that we cannot use _CMD from within Helm chart. I always thought of the _CMD pattern as one giving you false sense of security through obfuscation.

The _CMD is really a poor and not-really-secure solution, that was introduced because in "classic" deployments you have no other possibilities. But I think that is not that useful in K8S where you have other ways to get secret values: static Secrets, but also for example dynamically generated secretes (https://www.hashicorp.com/blog/dynamic-database-credentials-with-vault-and-kubernetes for example).

And it does not stop here. There might be cases where custom non-K8S integrated command might still be needed, but IMHO this would be much better to handle it via a custom sidecar that could share folder with secrets with Airflow and use https://airflow.apache.org/docs/apache-airflow/stable/security/secrets/secrets-backend/local-filesystem-secrets-backend.html  to read those. 

This has one huge advantage over _CMD pattern. Unlike the _CMD pattern, it is actually secure.

The sidecar could have the "long living" authentication token that would be inaccessible to the rest of the airflow, and it could only share the "short-living" tokens with Airflow. This is not really possible with _CMD pattern - this is merely an obfuscation, because no matter what the command run via _CMD must have somewhere the long-living authentication and determined user in Airflow could always find it if they really try. 

With the side-car pattern, it's not really possible by Airflow user to get hold of the long-living authentication credential. Very similar pattern is implemented in Kerberos support for Airlfow Charet, where only Kerberos sidecar has access to the long-living credentials (keytab) and it only shares the short-term token in shared volume that Airflow can read. See https://github.com/apache/airflow/blob/98c12d49f37f6879e3e9fd926853f57a15ab761b/chart/values.yaml#L270

I think we should encourage our users to use the other available options rather than rely on _CMD variables. 

If anything, I'd describe the side-car pattern in Helm documentation as "recommended" approach of dynamically retrieving credentials.

**Comment by mik-laj** [comment]
> I always thought of the _CMD pattern as one giving you false sense of security through obfuscation.

Environment variables in a containerized environment are sometimes perceived as insecure. In that case, we could use `_CMD`. See: https://docs.gitlab.com/charts/architecture/decisions.html#preference-of-secrets-in-initcontainer-over-environment

**Comment by mik-laj** [comment]
@potiuk  I also still don't know how you would want to integrate dynamic airflow configuration without using `_CMD`. LocalFilesystemBaackend only supports connection and variable, not airflow configuration.
`_CMD` environment variable is a glue that can connect the dynamic secrets mounted in volumes together with Airflow.

By convention Vault will inject these at the path `/vault/secrets/<secret name>`, the following snippet shows an example of this file.
```
databaseUrl: mysql://someuser:somepass@somehost:7777/somedb
```
If you want to use this value with Airflow, you have to pass it to Airflow. The easiest way is to set a _CMD variable that will point to this file.
```
export AIRFLOW__CORE__SQL_ALCHEMY_CONN_CMD="cat /vault/secrets/<secret name> | cut -d ":" -f 2-"
```
Do you know an alternative solution without writing custom code?

**Comment by potiuk** [comment]
Yeah. I think you are partially right @mik-laj, if you treat the _CMD as a "glue" to re-format  credentials provided in a different format from a more secure way of sharing credentials and the other way has no way to format it in the way it shoudl be provided, the _CMD pattern is quite OK. 

But .... 

(And i am not trying to make my point just trying to make it more clear where my thinking comes from and maybe see what you think).

The example with Vault is a good one - because it is an implementation of precisely the side-car pattern I explained (Vault injects the secrets exactly via side-car secret injector). So vault already encourages the use of the side-car pattern. And really the question is "where" the formatting glue should be added. From what I know, Vault actually support secret templates that provide very flexible way of injecting the secrets: https://www.vaultproject.io/docs/platform/k8s/injector#secret-templates - and the question here is who should adapt and provide the "glue". 

Should it be Airflow via _CMD pattern, or should it be secret 'injector'?

Note, I do not argue to remove the _CMD pattern altogether from Airflow. It is very useful in case of on-prem/custom deployments which have no other way to provide security. I am arguing that in the Helm/K8S world there are other ways how this can be achieved - where the custom code - if needed - shoudl be written "outside" of Airlfow. And my point is that we should discourage use of _CMD for that.

The case with "just reformatting" is fine, but I am afraid (and this is my whole point) that the _CMD pattern encourages people to write scripts that provide the "temporary" where the scripts themselves contain "hidden/obfuscated" credentials. For example I can imagine one can write a script to communicate with a secret manager where the "user/password" to the secret manager is embedded (and maybe obfuscated) in the script itself. This is the "false sense of security" that I am talking about. 

For me the example you wrote for re-formatting the secret is already the "custom code" that you need to write. And my argument is that either you do not need it (like vault + templating + maybe adding configuration to FileSystemBackend) or writing custom code (might be a bash script) in your "side-car-secret-injector". Is much better choice, which encourages much more secure patterns. Even if it is a bit more difficult.

BTW. I was not aware FileSystemBackend does not support configuration (other secret backends support it and it causes some pain in configuration class).

**Comment by mik-laj** [comment]
> From what I know, Vault actually support secret templates that provide very flexible way of injecting the secrets: https://www.vaultproject.io/docs/platform/k8s/injector#secret-templates - and the question here is who should adapt and provide the "glue".

We still need some glue, because the Vaultt Secret Template still only allows the secret to be saved to a file. You cannot set such a secret to an environment variable.
```
export AIRFLOW__CORE__SQL_ALCHEMY_CONN_CMD="cat /vault/secrets/<secret name>"
```
> For example I can imagine one can write a script to communicate with a secret manager where the "user/password" to the secret manager is embedded (and maybe obfuscated) in the script itself.

There are many ways for the user to badly manage their passwords, whether they are read from a script or otherwise. Rather, we will not be able to protect against it, but we should show how to do it well, e.g. support passwords managed by the Vault Injector in our Helm Charrt. 

> maybe adding configuration to FileSystemBackend

This can make the configuration even more difficult, because it means that we will have another Airflow configuration file in addition to `airflow.cfg`. Also, I'm not sure if this will actually simplify our configurations when we have to figure out how to write a template that generates the configurations to a single file that will be read by the secret backend and set up secret backend instead of just setting one envariable variable that contains one bash command - `cat /vault/secrets/<secret name>`.

> BTW. I was not aware FileSystemBackend does not support configuration (other secret backends support it and it causes some pain in configuration class).

Configuration support was added later and not every backend has to support this. We only added it when it made sense, and here it was minimal when we already had the airflow.cfg file with the configuration.

**Comment by mik-laj** [comment]
@jpyen  Can you tell me a little more about your use case? Why is it important to you? Do you integrate with other tools?

**Comment by mik-laj** [comment]
@jedcunningham  WDYT?

**Comment by potiuk** [comment]
Yeah. I'd also love to hear what others think. I would be in favour (even if it is not currently supported) to find something that is more Helm-native without having to fallback to custom _CMD glue. Maybe we can find a better approach that is more natural to use and encourages secure isolation between long term and short-term credentials.

For me any solution which directly promotes using side-cars and where short-term credentials only are shared with the airflow workers is a step in good direction.

**Comment by jedcunningham** [comment]
Another option, granted with a little more complexity, is to have an init container be fully responsible for building a complete airflow.cfg, which Airflow then uses via the shared emptyDir. This would nicely separate the config stuff and Airflow, and hopefully is flexible enough to support other secret approaches too.

Said another way, have the init container take the "base" airflow.cfg and inject (in the case of vault) stuff from /vault/secrets/* into a combined airflow.cfg.

**Comment by potiuk** [comment]
How about simply using valueFrom option. Seems that you could pass name of the secret and key and set it as environment variable:

```
    env:
      - name: SECRET_USERNAME
        valueFrom:
          secretKeyRef:
            name: mysecret
            key: username
      - name: SECRET_PASSWORD
        valueFrom:
          secretKeyRef:
            name: mysecret
            key: password

```

In the case from @mik-laj it would be:

```
env:
  - name: AIRFLOW__CORE__SQL_ALCHEMY_CONN_CMD
    valueFrom:
       secretKeyRef:
           name: vault-secret-name
           key: databaseUrl
```

Sounds pretty cool and flexible, I am not sure if we can already do it in the current chart easily, but I think that looks like pretty good native solution - with a bit less flexibility but without promoting the _CMD pattern.

**Comment by mik-laj** [comment]
> using valueFrom option.

It uses the native secret backend provided by Kubernetes. Unfortunately, he is not considered to be very safe and is quite limited, e.g. it does not support dynamic secrets.  Many corporate users prefer solutions that meet more stringent requirements, such as Hashicorp Vault.  

Unfortunately, This is a limitation of Kubernetes, which has decided to be an all-in-one solution and third-party solutions are supported in a rather limited way.  This design problem was solved by Hashicorp Nomad, which does not provide any built-in manager secret, and if you need to, you must use another tool that turns out to be provided by the same company - Hashicorp Vault. For details see, https://learn.hashicorp.com/tutorials/nomad/vault-postgres?in=nomad/integrate-vault

The use of environment variables is also quite limiting. If the user would like to change the secret value after starting the container, the secret value will not be updated in the container.  There are only third-party solutions that can then force the container to restart so that the variables can be updated.

> Environment variables are not updated after a secret update
> If a container already consumes a Secret in an environment variable, a Secret update will not be seen by the container unless it is restarted. There are third party solutions for triggering restarts when secrets change.

https://kubernetes.io/docs/concepts/configuration/secret/#environment-variables-are-not-updated-after-a-secret-update

**Comment by mik-laj** [comment]
>  I am not sure if we can already do it in the current chart easily, 

This is the currently used supported solution by Helm Chart, but we want to add support for non-native secret managers.
https://github.com/apache/airflow/blob/2b7c59619b7dd6fd5031745ade7756466456f803/chart/values.yaml#L238-L240
https://github.com/apache/airflow/blob/36651becaf9c706452e987c4050f45994ac2c0de/chart/templates/_helpers.yaml#L45-L54

**Comment by thesuperzapper** [comment]
Those watching might be interested in how the ["user community" chart](https://github.com/airflow-helm/charts/tree/main/charts/airflow) (not the "official" one) manages database connection strings.

The "user community" chart exposes configs for "parts" of the connection string, rather than making the user provide the entire string. Most importantly, the user only has to define a Kubernetes Secret which contains their database password with the [`externalDatabase.passwordSecret`](https://github.com/airflow-helm/charts/blob/3721cfd433b5a210361def6afd63b725f9ac4044/charts/airflow/values.yaml#L1627-L1629) value. You can see the other `externalDatabase.*` values [here](https://github.com/airflow-helm/charts/blob/3721cfd433b5a210361def6afd63b725f9ac4044/charts/airflow/values.yaml#L1602-L1641).

This is enabled by internally using the [`_CMD` variables](https://github.com/airflow-helm/charts/blob/3721cfd433b5a210361def6afd63b725f9ac4044/charts/airflow/templates/config/secret-config-envs.yaml#L17-L64) so the "password" secret can be used, even though we don't know its value at chart install time.
